jane/doc/parallelism-tutorial.md

```diff
@@ -0,0 +1,1030 @@
+`Parallel.Sequence` module makes it nearly trivial:
+
+```ocaml
+let add_many_par par arr =
+  let seq = Parallel.Sequence.of_iarray arr in
+  Parallel.Sequence.reduce par seq ~f:(fun a b -> a + b)
+  |> Option.value ~default:0
+```
+
+We first need to convert from `iarray` to `Parallel.Sequence.t`, a general
+sequence type supporting a rich selection of parallel operations. Among them is
+a parallel `reduce`, which operates much like an unordered `fold`.
+
+Just like `Parallel.fork_join2`, we can use parallel sequences in a nested
+manner. Suppose that rather than a binary tree we have an n-ary tree:
+
+```ocaml
+module Tree = struct
+  type 'a t =
+    | Leaf of 'a
+    | Nodes of 'a t iarray
+end
+```
+<!--
+Importantly, we're using `iarray`s rather than the usual mutable `array`s
+here—otherwise there would be little hope of doing anything in parallel. The
+sequential code isn't much different from [before]:
+-->
+
+The sequential code isn't much different from [before]:
+
+[before]: #code-average
+
+```ocaml
+let average tree =
+  let rec total tree =
+    match tree with
+    | Tree.Leaf x -> ~total:(Thing.price x), ~count:1
+    | Tree.Nodes arr ->
+      let totals_and_counts =
+        Iarray.map arr ~f:(fun subtree -> total subtree)
+      in
+      Iarray.fold
+        totals_and_counts
+        ~init:(~total:0.0, ~count:0)
+        ~f:(fun (~total:total_acc, ~count:count_acc) (~total, ~count) ->
+          ~total:(total +. total_acc), ~count:(count + count_acc))
+  in
+  let ~total, ~count = total tree in
+  total /. (count |> Float.of_int)
+;;
+```
+
+Naturally, we now use `Iarray.map` to recurse on subnodes and `Iarray.fold` to
+combine the results. To parallelize, we still need the
+`Parallel.Sequence.of_iarray`, but now `Parallel.Sequence.fold'` combines
+the `map` and `reduce` operations:
+
+```ocaml
+let average_par (par : Parallel.t) tree =
+  let rec (total @ portable) par tree =
+    match tree with
+    | Tree.Leaf x -> ~total:(Thing.price x), ~count:1
+    | Tree.Node arr ->
+      let seq = Parallel.Sequence.of_iarray arr in
+      Parallel.Sequence.fold'
+        par
+        seq
+        ~f:(fun par subtree -> total par subtree)
+        ~init:(~total:0.0, ~count:0)
+        ~combine:(fun _par (~total, ~count) (~total:total2, ~count:count2) ->
+          ~total:(total +. total2), ~count:(count + count2))
+      [@nontail]
+  in
+  let ~total, ~count = total par tree in
+  total /. (count |> Float.of_int)
+```
+
+(We use `fold'` rather than `fold` to get the version that passes `par` down
+into our `f` and `combine` functions.)
+
+Another approach would be to convert the entire tree into a
+`Parallel.Sequence.t` at once rather than rely on nested parallelism. This is
+possible using `Parallel.Sequence.unfold`, though it does take more work,
+requiring a definition of an unfolding state with an operation that splits it in
+half.
+
+**[I made a solution for this but unless there's something I missed that would
```

Comment:
Hard for me to see exactly where this is going, so I don't have much of an opinion here. I do like the idea of generalizing the initial example to something a little harder.